content/blog/2024-07-07-aws-to-contribute-efficiency-improvements-for-ValKey-8.md

```diff
@@ -0,0 +1,67 @@
+transactions, more powerful shards simplify key distribution when designing the application by allowing a higher
+concentration of slots in a single shard. Finally, higher-capacity shards handling requests for a larger number
+of slots are more effective in load balancing resources during arbitrary increases in load on random slots.
+This reduces the need for costly over-provisioning.
+
+### Performance Without Compromising Simplicity
+
+Implementing multi-threading  can be a complex task. Over the years, contributors have been careful to maintain Redis’ and now Valkey’s simplicity . This ensures an API  that can continuously evolve without the need to use complex synchronization and avoid race conditions. 
+
+We designed our new multi-threading approach to delight Valkey users with intuitive simplicity. It utilizes a minimal number of synchronization mechanisms and keeps Valkey command execution single-threaded, simple, and primed for future enhancements.
+
+![io_threads high level design](/assets/media/pictures/io_threads.png)
+
+### High Level Design 
+The above diagram depicts the io_threads implementation from a high-level perspective. Io_threads are stateless worker threads that receive jobs to execute from the main thread. In Valkey 8, a job can involve reading and parsing a command from a client, writing back responses to the client, polling for IO events on TCP connections, or de-allocating memory. This leaves the main thread with more time to execute  commands. 
+
+The main thread orchestrates all the jobs spawned to the io_threads, ensuring that no race conditions occur. Io_threads can be easily added and removed by the main thread based on the current load to ensure efficient utilization of the underlying hardware. Despite the dynamic nature of io_threads, the main thread attempts to maintain thread affinity, ensuring that the same io_thread will handle IO for the same client to improve memory access locality. 
+
+Before executing commands, the main thread performs a new procedure, prefetch-commands-keys, which aims to reduce the number of external memory accesses needed when executing the commands on the main dictionary. A detailed explanation of the technique used in that procedure will be described in our next blog. 
+
+Socket polling system calls, such as epoll_wait, are expensive procedures. When executed solely by the main thread, epoll_wait consumes more than 20 percent of the time. Therefore, we decided to offload epoll_wait execution to the io_threads in the following way: to avoid race conditions, at any time, at most one thread, either an io_thread or the main thread, is executing epoll_wait. Io_threads never sleep on epoll, and whenever there are pending IO operations or commands to be executed, epoll_wait calls are scheduled to the io_threads by the main thread. In all other cases, the main thread executes the epoll_wait with the waiting time as in the original Valkey implementation 
```

Comment:
```suggestion
Socket polling system calls, such as epoll_wait, are expensive procedures.
When executed solely by the main thread, epoll_wait consumes more than 20 percent of the CPU time of the process.
Therefore, we offload `epoll_wait` execution to the IO thread when necessary by scheduling an epoll job from the main thread to an IO thread.
To avoid race conditions, the main thread will no longer execute `epoll_wait` until the poll job has completed, ensuring that only one thread is executing the `epoll_wait` at a given time.
```